File: sdks/oracles/ts/src/instructions/er.ts
/**
 * Ephemeral-rollup session builders (Ix 24–26): delegate / commit / undelegate.
 *
 * Short form writes only Kassandra's `ErSession` companion PDA. Production
 * MagicBlock CPI remaining-accounts are appended by the caller when present.
 */
import { Address, TransactionInstruction } from "@solana/web3.js";

import { Ix, KASSANDRA_PROGRAM_ID, SYSTEM_PROGRAM_ID } from "../constants.js";
import * as pda from "../pda.js";
import type { AddressInput } from "../pda.js";
import { addr, pubkeyBytes, ro, u32LE, u64LE, w, withDisc } from "./payload.js";

const ZERO32 = new Uint8Array(32);

// ---------------------------------------------------------------------------
// DelegateOracle (Ix=24) — processor/delegate_oracle.rs
// Accounts (short): 0 oracle(w) 1 er_session(w) 2 payer(signer,w) 3 system(ro).
// Payload (44): nonce u64 LE ++ commit_frequency_ms u32 LE ++ validator [32].
// `commit_frequency_ms == 0` → MagicBlock default 30s; validator all-zero → none.
// ---------------------------------------------------------------------------
export interface DelegateOracleArgs {
  oracle: AddressInput;
  payer: AddressInput;
  /** Oracle PDA nonce (u64), same as CreateOracle. */
  nonce: bigint | number;
  /** Commit cadence in ms. `0` (default) lets the program stamp 30_000. */
  commitFrequencyMs?: number;
  /** ER validator pubkey. Omitted / all-zero → Magic Router default. */
  validator?: AddressInput;
  programId?: Address;
}

export async function delegateOracle(args: DelegateOracleArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? KASSANDRA_PROGRAM_ID;
  const oracle = addr(args.oracle);
  const session = await pda.erSession(oracle, programId);
  const validator = args.validator ? pubkeyBytes(args.validator) : ZERO32;
  const data = withDisc(
    Ix.DelegateOracle,
    u64LE(args.nonce),
    u32LE(args.commitFrequencyMs ?? 0),
    validator,
  );
  return new TransactionInstruction({
    programId,
    keys: [w(oracle), w(session.address), w(addr(args.payer), true), ro(SYSTEM_PROGRAM_ID)],
    data,
  });
}

// ---------------------------------------------------------------------------
// CommitOracle (Ix=25) — processor/commit_oracle.rs
// Accounts: 0 oracle(w) 1 er_session(w). Payload: empty.
// ---------------------------------------------------------------------------
export interface CommitOracleArgs {
  oracle: AddressInput;
  programId?: Address;
}

export async function commitOracle(args: CommitOracleArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? KASSANDRA_PROGRAM_ID;
  const oracle = addr(args.oracle);
  const session = await pda.erSession(oracle, programId);
  return new TransactionInstruction({
    programId,
    keys: [w(oracle), w(session.address)],
    data: withDisc(Ix.CommitOracle),
  });
}

// ---------------------------------------------------------------------------
// UndelegateOracle (Ix=26) — processor/undelegate_oracle.rs
// Accounts: 0 oracle(w) 1 er_session(w). Payload: empty.
// ---------------------------------------------------------------------------
export interface UndelegateOracleArgs {
  oracle: AddressInput;
  programId?: Address;
}

export async function undelegateOracle(args: UndelegateOracleArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? KASSANDRA_PROGRAM_ID;
  const oracle = addr(args.oracle);
  const session = await pda.erSession(oracle, programId);
  return new TransactionInstruction({
    programId,
    keys: [w(oracle), w(session.address)],
    data: withDisc(Ix.UndelegateOracle),
  });
}

File: sdks/oracles/rust/src/ix/er.rs
//! Ephemeral-rollup session instruction builders (Ix 24–26).

use kassandra_oracles_program::instruction::Ix;
use solana_instruction::{AccountMeta, Instruction};
use solana_pubkey::Pubkey;

use super::build;
use crate::SYSTEM_PROGRAM_ID;

/// `DelegateOracle` (Ix 24) — short form (Kassandra `ErSession` only).
/// Payload: `nonce u64 LE ++ commit_frequency_ms u32 LE ++ validator [u8;32]`
/// (`commit_frequency_ms == 0` → MagicBlock default 30s; validator all-zero → none).
pub fn delegate_oracle(
    program_id: &Pubkey,
    oracle: Pubkey,
    er_session: Pubkey,
    payer: Pubkey,
    nonce: u64,
    commit_frequency_ms: u32,
    validator: &Pubkey,
) -> Instruction {
    let mut data = Vec::with_capacity(1 + 44);
    data.push(Ix::DelegateOracle as u8);
    data.extend_from_slice(&nonce.to_le_bytes());
    data.extend_from_slice(&commit_frequency_ms.to_le_bytes());
    data.extend_from_slice(validator.as_ref());
    build(
        program_id,
        vec![
            AccountMeta::new(oracle, false),
            AccountMeta::new(er_session, false),
            AccountMeta::new(payer, true),
            AccountMeta::new_readonly(SYSTEM_PROGRAM_ID, false),
        ],
        data,
    )
}

/// `CommitOracle` (Ix 25).
pub fn commit_oracle(program_id: &Pubkey, oracle: Pubkey, er_session: Pubkey) -> Instruction {
    build(
        program_id,
        vec![
            AccountMeta::new(oracle, false),
            AccountMeta::new(er_session, false),
        ],
        vec![Ix::CommitOracle as u8],
    )
}

/// `UndelegateOracle` (Ix 26).
pub fn undelegate_oracle(program_id: &Pubkey, oracle: Pubkey, er_session: Pubkey) -> Instruction {
    build(
        program_id,
        vec![
            AccountMeta::new(oracle, false),
            AccountMeta::new(er_session, false),
        ],
        vec![Ix::UndelegateOracle as u8],
    )
}
